File: riveredge-backend/src/core/models/approval_process.py
"""
审批流程模型模块

定义审批流程数据模型，用于审批流程管理。
"""

from tortoise import fields
from typing import Optional, Dict, Any
from .base import BaseModel


class ApprovalProcess(BaseModel):
    """
    审批流程模型
    
    用于定义和管理组织内的审批流程，支持多组织隔离。
    
    注意：继承自 BaseModel，自动包含 uuid、tenant_id、created_at、updated_at 字段。
    """
    id = fields.IntField(pk=True, description="审批流程ID（主键，自增ID，内部使用）")
    # uuid 字段由 BaseModel 提供
    # tenant_id 字段由 BaseModel 提供
    
    name = fields.CharField(max_length=100, description="流程名称")
    code = fields.CharField(max_length=50, description="流程代码（唯一，用于程序识别）")
    description = fields.TextField(null=True, description="流程描述")
    
    # 流程配置
    nodes = fields.JSONField(description="流程节点配置（JSON格式，ProFlow 设计）")
    config = fields.JSONField(description="流程配置（JSON格式）")
    
    # 流程状态
    is_active = fields.BooleanField(default=True, description="是否启用")

    version = fields.IntField(default=1, description="流程版本（发布后递增）")
    published_version = fields.IntField(default=1, description="已发布版本号")
    draft_nodes = fields.JSONField(null=True, description="草稿节点（未发布编辑）")
    
    # 软删除字段
    deleted_at = fields.DatetimeField(null=True, description="删除时间（软删除）")
    
    class Meta:
        """
        模型元数据
        """
        table = "core_approval_processes"
        indexes = [
            ("tenant_id", "code"),  # 唯一索引
            ("uuid",),
            ("code",),
            ("is_active",),
            ("created_at",),
        ]
        unique_together = [("tenant_id", "code")]
    
    def __str__(self):
        """字符串表示"""
        return f"{self.name} ({self.code})"

File: riveredge-frontend/src/services/approvalProcess.ts
/**
 * 审批流程管理服务
 * 
 * 提供审批流程的 CRUD 操作。
 * 注意：所有 API 自动过滤当前组织的审批流程
 */

import { apiRequest } from './api';

export interface ApprovalProcess {
  uuid: string;
  tenant_id: number;
  name: string;
  code: string;
  description?: string;
  nodes: Record<string, any>;
  config: Record<string, any>;
  draft_nodes?: Record<string, any> | null;
  version?: number;
  published_version?: number;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface ApprovalProcessListParams {
  skip?: number;
  limit?: number;
  is_active?: boolean;
  /** 配置中心审核设置：返回全部审核开关流程（不受已安装应用过滤） */
  for_audit_config?: boolean;
}

export interface CreateApprovalProcessData {
  name: string;
  code: string;
  description?: string;
  nodes: Record<string, any>;
  config: Record<string, any>;
  is_active?: boolean;
}

export interface UpdateApprovalProcessData {
  name?: string;
  description?: string;
  nodes?: Record<string, any>;
  config?: Record<string, any>;
  is_active?: boolean;
}

/**
 * 获取审批流程列表
 */
export async function getApprovalProcessList(params?: ApprovalProcessListParams): Promise<ApprovalProcess[]> {
  return apiRequest<ApprovalProcess[]>('/core/approval-processes', {
    params,
  });
}

/**
 * 获取审批流程详情
 */
export async function getApprovalProcessByUuid(approvalProcessUuid: string): Promise<ApprovalProcess> {
  return apiRequest<ApprovalProcess>(`/core/approval-processes/${approvalProcessUuid}`);
}

/**
 * 创建审批流程
 */
export async function createApprovalProcess(data: CreateApprovalProcessData): Promise<ApprovalProcess> {
  return apiRequest<ApprovalProcess>('/core/approval-processes', {
    method: 'POST',
    data,
  });
}

export async function getConditionFields(entityType: string): Promise<{ fields: Array<{ field: string; label: string; type: string; operators: string[] }> }> {
  return apiRequest(`/core/approval-processes/condition-fields`, { params: { entity_type: entityType } });
}

/**
 * 更新审批流程
 */
export async function updateApprovalProcess(approvalProcessUuid: string, data: UpdateApprovalProcessData): Promise<ApprovalProcess> {
  return apiRequest<ApprovalProcess>(`/core/approval-processes/${approvalProcessUuid}`, {
    method: 'PUT',
    data,
  });
}

export async function publishApprovalProcess(approvalProcessUuid: string): Promise<ApprovalProcess> {
  return apiRequest<ApprovalProcess>(`/core/approval-processes/${approvalProcessUuid}/publish`, {
    method: 'POST',
  });
}

/**
 * 删除审批流程
 */
export async function deleteApprovalProcess(approvalProcessUuid: string): Promise<void> {
  return apiRequest<void>(`/core/approval-processes/${approvalProcessUuid}`, {
    method: 'DELETE',
  });
}

